README.md
## openplotter-moitessier

OpenPlotter app to manage the Moitessier HAT

### Installing

#### For production

Install [openplotter-settings](https://github.com/openplotter/openplotter-settings) for **production** and just install this app from *OpenPlotter Apps* tab.

#### For development

Install [openplotter-settings](https://github.com/openplotter/openplotter-settings) for **development**.

Clone the repository:

`git clone https://github.com/openplotter/openplotter-moitessier`

Make your changes and create the package:

```
cd openplotter-moitessier
dpkg-buildpackage -b
```

Install the package:

```
cd ..
sudo dpkg -i openplotter-moitessier_x.x.x-xxx_all.deb
```

Run post-installation script:

`sudo moitessierPostInstall`

Run:

`openplotter-moitessier`

Make your changes and repeat package, installation and post-installation steps to test. Pull request your changes to github and we will check and add them to the next version of the [Debian package](https://launchpad.net/~openplotter/+archive/ubuntu/openplotter).

### Documentation

https://openplotter.readthedocs.io

### Support

http://forum.openmarine.net/forumdisplay.php?fid=1


debian/changelog
openplotter-moitessier (2.4.2-stable) bionic; urgency=medium

  * Update drivers
  * Update translations
  * Fix error in ubuntu for raspberry
  * Add warning if hat is not attached
  * remove i2cdetect

 -- Sailoog <info@sailoog.com>  Thu, 03 Dec 2020 21:22:45 +0100

openplotter-moitessier (2.4.0-stable) bionic; urgency=medium

  * Remove dtoverlay at install
  * Update translations 

 -- Sailoog <info@sailoog.com>  Mon, 02 Nov 2020 17:49:28 +0100

openplotter-moitessier (2.3.0-stable) bionic; urgency=medium

  * Add GPIO management
  * Update drivers
  * Update translations

 -- Sailoog <info@sailoog.com>  Mon, 19 Oct 2020 18:03:13 +0200

openplotter-moitessier (2.2.0-stable) bionic; urgency=medium

  * Update configuration checking process to add OpenCPN Signal K connection
  * Update Moitesseier HAT drivers
  * Update translations

 -- Sailoog <info@sailoog.com>  Sat, 08 Aug 2020 11:53:47 +0200

openplotter-moitessier (2.1.2-stable) bionic; urgency=medium

  * Update to Moitessier HAT driver 1.8.1

 -- Sailoog <info@sailoog.com>  Tue, 17 Mar 2020 18:38:44 +0100

openplotter-moitessier (2.1.1-stable) bionic; urgency=medium

  * Update Moitessier HAT drivers
  * Update translations

 -- Sailoog <info@sailoog.com>  Mon, 09 Mar 2020 19:38:15 +0100

openplotter-moitessier (2.1.0-stable) bionic; urgency=medium

  * Initial release

 -- Sailoog <info@sailoog.com>  Thu, 27 Feb 2020 19:43:39 +0100

openplotter-moitessier (2.0.7-stable) bionic; urgency=medium

  * Candidate release

 -- Sailoog <info@sailoog.com>  Sat, 15 Feb 2020 13:41:02 +0100


openplotterMoitessier/startup.py
#!/usr/bin/env python3

# This file is part of Openplotter.
# Copyright (C) 2019 by Sailoog <https://github.com/openplotter/openplotter-moitessier>
#                  
# Openplotter is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# any later version.
# Openplotter is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Openplotter. If not, see <http://www.gnu.org/licenses/>.

import os, subprocess, sys
from openplotterSettings import language


class Start():
	def __init__(self, conf, currentLanguage):
		self.initialMessage = ''

	def start(self):
		green = '' 
		black = '' 
		red = '' 

		return {'green': green,'black': black,'red': red}


class Check():
	def __init__(self, conf, currentLanguage):
		self.conf = conf
		currentdir = os.path.dirname(os.path.abspath(__file__))
		language.Language(currentdir,'openplotter-moitessier',currentLanguage)
		
		self.initialMessage = _('Checking Moitessier HAT...')

	def check(self):
		green = '' 
		black = '' 
		red = '' 

		attached = False
		try:
			out = subprocess.check_output(['more','product'],cwd='/proc/device-tree/hat').decode(sys.stdin.encoding)
		except: black =_('not attached')
		else:
			if not 'Moitessier' in out: black =_('not attached')
			else: 
				green =_('attached')
				attached = True

		try:
			out = subprocess.check_output('ls /dev/i2c*', shell=True).decode(sys.stdin.encoding)
			if '/dev/i2c-0' in out: red = _('Your Raspberry Pi is too old!')
			if '/dev/i2c-1' in out:
				txt = _('I2C enabled')
				if not black: black = txt
				else: black += ' | '+txt
		except:
			txt = _('I2C is disabled. Please enable I2C interface in Preferences -> Raspberry Pi configuration -> Interfaces.')
			if not red: red = txt
			else: red += '\n'+txt

		spi_bcm2835 = subprocess.check_output('lsmod').decode(sys.stdin.encoding)
		if 'spi_bcm2835' in spi_bcm2835:
			txt = _('SPI enabled')
			if not black: black = txt
			else: black += ' | '+txt
		else:
			txt = _('SPI is disabled. Please enable SPI interface in Preferences -> Raspberry Pi configuration -> Interfaces.')
			if not red: red = txt
			else: red += '\n'+txt

		modulesPath = self.conf.home+'/moitessier/modules'
		if not os.path.exists(modulesPath):
			txt = _('Moitessier HAT driver is not installed.')
			if not red: red = txt
			else: red += '\n'+txt
		else:
			txt = _('driver installed')
			if not black: black = txt
			else: black += ' | '+txt
			if not attached:
				txt = _('Moitessier HAT is not attached, uninstall the driver if you are not using this HAT')
				if not red: red = txt
				else: red += '\n'+txt

			packages = os.listdir(modulesPath)
			kernel = subprocess.check_output(['uname','-r']).decode(sys.stdin.encoding)
			kernel = kernel.replace('\n','')
			supported = False
			for i in packages:
				if 'moitessier_'+kernel+'.ko' == i: supported = True
			if not supported:
				txt = _('The installed driver does not support the current kernel version, go to Moitessier HAT app to update it.')
				if not red: red = txt
				else: red += '\n'+txt

		return {'green': green,'black': black,'red': red}


openplotterMoitessier/moitessierPostInstall.py
#!/usr/bin/env python3

# This file is part of Openplotter.
# Copyright (C) 2019 by Sailoog <https://github.com/openplotter/openplotter-moitessier>
#
# Openplotter is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# any later version.
# Openplotter is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Openplotter. If not, see <http://www.gnu.org/licenses/>.
import os, shutil, subprocess
from openplotterSettings import conf
from openplotterSettings import language
from .version import version

def main():
	conf2 = conf.Conf()
	currentdir = os.path.dirname(os.path.abspath(__file__))
	currentLanguage = conf2.get('GENERAL', 'lang')
	language.Language(currentdir,'openplotter-moitessier',currentLanguage)


	print(_('Copying driver packages...'))
	try:
		confFolder = conf2.conf_folder+'/moitessier'
		if not os.path.exists(confFolder): os.mkdir(confFolder)

		packageFolder = currentdir+'/packages'
		if os.path.exists(packageFolder):
			tmp = os.listdir(packageFolder)
			for i in tmp:
				shutil.copy(packageFolder+'/'+i, confFolder)

		subprocess.call(['chown', '-R', conf2.user, confFolder])

		print(_('DONE'))
	except Exception as e: print(_('FAILED: ')+str(e))

	print(_('Setting version...'))
	try:
		conf2.set('APPS', 'moitessier', version)
		print(_('DONE'))
	except Exception as e: print(_('FAILED: ')+str(e))
	
if __name__ == '__main__':
	main()

openplotterMoitessier/moitessierPreUninstall.py
#!/usr/bin/env python3

# This file is part of Openplotter.
# Copyright (C) 2019 by Sailoog <https://github.com/openplotter/openplotter-moitessier>
#                  
# Openplotter is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# any later version.
# Openplotter is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Openplotter. If not, see <http://www.gnu.org/licenses/>.

import os, shutil, subprocess
from openplotterSettings import conf
from openplotterSettings import language

def main():
	conf2 = conf.Conf()
	currentdir = os.path.dirname(os.path.abspath(__file__))
	currentLanguage = conf2.get('GENERAL', 'lang')
	language.Language(currentdir,'openplotter-moitessier',currentLanguage)


	print(_('Removing drivers packages...'))
	try:
		confFolder = conf2.conf_folder+'/moitessier'
		shutil.rmtree(confFolder)

		print(_('DONE'))
	except Exception as e: print(_('FAILED: ')+str(e))

	print(_('Uninstalling drivers...'))
	try:
		subprocess.call(['dpkg', '-r', 'moitessier'])

		config = '/boot/config.txt'
		boot = '/boot'
		try: file = open(config, 'r')
		except:
			config = '/boot/firmware/config.txt'
			boot = '/boot/firmware'
			file = open(config, 'r')
		file1 = open('config.txt', 'w')
		while True:
			line = file.readline()
			if not line: break
			if 'dtoverlay=i2c-gpio,i2c_gpio_sda=2,i2c_gpio_scl=3,bus=3' in line: pass
			else: file1.write(line)
		file.close()
		file1.close()
		if os.system('diff config.txt '+config+' > /dev/null'): os.system('mv config.txt '+boot)
		else: os.system('rm -f config.txt')

		print(_('DONE'))
	except Exception as e: print(_('FAILED: ')+str(e))

	print(_('Removing version...'))
	try:
		conf2.set('APPS', 'moitessier', '')
		print(_('DONE'))
	except Exception as e: print(_('FAILED: ')+str(e))
	
if __name__ == '__main__':
	main()

openplotterMoitessier/gpio.py
#!/usr/bin/env python3

# This file is part of Openplotter.
# Copyright (C) 2019 by Sailoog <https://github.com/openplotter/openplotter-moitessier>
#                     
# Openplotter is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# any later version.
# Openplotter is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Openplotter. If not, see <http://www.gnu.org/licenses/>.
import subprocess, sys

class Gpio:
	def __init__(self,conf):
		self.conf = conf
		self.used = [] # {'app':'xxx', 'id':'xxx', 'physical':'n'}

	def usedGpios(self):
		try:
			out = subprocess.check_output(['more','product'],cwd='/proc/device-tree/hat').decode(sys.stdin.encoding)
		except:pass
		else:
			if 'Moitessier' in out:
				self.used.append({'app':'Moitessier', 'id':'power', 'physical':'1'})
				self.used.append({'app':'Moitessier', 'id':'power', 'physical':'2'})
				self.used.append({'app':'Moitessier', 'id':'Press/temp/heading', 'physical':'3'})
				self.used.append({'app':'Moitessier', 'id':'power', 'physical':'4'})
				self.used.append({'app':'Moitessier', 'id':'Press/temp/heading', 'physical':'5'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'6'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'9'})
				self.used.append({'app':'Moitessier', 'id':'output', 'physical':'11'})
				self.used.append({'app':'Moitessier', 'id':'output', 'physical':'12'})
				self.used.append({'app':'Moitessier', 'id':'input', 'physical':'13'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'14'})
				self.used.append({'app':'Moitessier', 'id':'output', 'physical':'15'})
				self.used.append({'app':'Moitessier', 'id':'input', 'physical':'16'})
				self.used.append({'app':'Moitessier', 'id':'power', 'physical':'17'})
				self.used.append({'app':'Moitessier', 'id':'output', 'physical':'18'})
				self.used.append({'app':'Moitessier', 'id':'GNSS/AIS', 'physical':'19'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'20'})
				self.used.append({'app':'Moitessier', 'id':'GNSS/AIS', 'physical':'21'})
				self.used.append({'app':'Moitessier', 'id':'GNSS/AIS', 'physical':'23'})
				self.used.append({'app':'Moitessier', 'id':'GNSS/AIS', 'physical':'24'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'25'})
				self.used.append({'app':'Moitessier', 'id':'EEPROM', 'physical':'27'})
				self.used.append({'app':'Moitessier', 'id':'EEPROM', 'physical':'28'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'30'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'34'})
				self.used.append({'app':'Moitessier', 'id':'ground', 'physical':'39'})
		return self.used